acceptance/github_asset_fetcher.go

```diff
@@ -0,0 +1,420 @@
+package acceptance
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/Masterminds/semver"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-github/v30/github"
+	"github.com/pkg/errors"
+
+	"github.com/buildpacks/pack/internal/blob"
+	"github.com/buildpacks/pack/internal/config"
+	ilogging "github.com/buildpacks/pack/internal/logging"
+)
+
+const (
+	assetCacheDir      = "test-assets-cache"
+	assetCacheManifest = "github.json"
+)
+
+type GithubAssetFetcher struct {
+	ctx          context.Context
+	logger       *ilogging.LogWithWriters
+	githubClient *github.Client
+	cacheDir     string
+}
+
+type assetCache map[string]map[string]cachedRepo
+type cachedRepo struct {
+	Assets   cachedAssets
+	Sources  cachedSources
+	Versions cachedVersions
+}
+type cachedAssets map[string][]string
+type cachedSources map[string]string
+type cachedVersions map[string]string
+
+func NewGithubAssetFetcher() (*GithubAssetFetcher, error) {
+	packHome, err := config.PackHome()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting pack home")
+	}
+
+	return &GithubAssetFetcher{
+		ctx:          context.TODO(),
+		logger:       ilogging.NewLogWithWriters(os.Stdout, os.Stderr), // TODO: not sure if this is what we really want
+		githubClient: github.NewClient(nil),
+		cacheDir:     filepath.Join(packHome, assetCacheDir),
+	}, nil
+}
+
+func (f *GithubAssetFetcher) FetchReleaseAsset(owner, repo, version string, expr *regexp.Regexp, extract bool) (string, error) {
+	if destPath, _ := f.cachedAsset(owner, repo, version, expr); destPath != "" {
+		fmt.Printf("found %s in cache for %s/%s %s\n", destPath, owner, repo, version)
+		return destPath, nil
+	}
+
+	release, _, err := f.githubClient.Repositories.GetReleaseByTag(f.ctx, owner, repo, version)
+	if err != nil {
+		return "", errors.Wrap(err, "getting release")
+	}
+
+	var desiredAsset *github.ReleaseAsset
+	for _, asset := range release.Assets {
+		if expr.MatchString(*asset.Name) {
+			desiredAsset = asset
+			break
+		}
+	}
+	if desiredAsset == nil {
+		return "", fmt.Errorf("could not find asset matching expression %s", expr.String())
+	}
+
+	var returnPath string
+	if extract && strings.Contains(*desiredAsset.Name, ".tgz") {
```

Comment:
This is a bit awkward. I am doing this in one big `if/else` block so that we can write once to the cache before returning.